gef.py

```diff
@@ -621,18 +621,99 @@ def __str__(self):
     def is_valid(self):
         return "(bad)" not in self.mnemonic
 
+class MallocStateStruct(object):
+    def __init__(self, addr):
+        self.addr = addr
+        self.num_fastbins = 11 if is_x86_32() else 10
+        self.num_bins = 254
+        self.size_t = cached_lookup_type("size_t")
+        if not self.size_t:
+            ptr_type = "unsigned long" if current_arch.ptrsize == 8 else "unsigned int"
+            self.size_t = cached_lookup_type(ptr_type)
+
+    @property
+    def fastbins_addr(self):
+        return self.addr + 8 + current_arch.ptrsize
+    @property
+    def top_addr(self):
+        return self.fastbins_addr + self.num_fastbins*current_arch.ptrsize
+    @property
+    def last_remainder_addr(self):
+        return self.top_addr + current_arch.ptrsize
+    @property
+    def bins_addr(self):
+        return self.last_remainder_addr + current_arch.ptrsize
+    @property
+    def next_addr(self):
+        return self.bins_addr + self.num_bins*current_arch.ptrsize + 16
+    @property
+    def next_free_addr(self):
+        return self.next_addr + current_arch.ptrsize
+    @property
+    def system_mem_addr(self):
+        return self.next_free_addr + current_arch.ptrsize*2
+
+    @property
+    def fastbinsY(self):
+        return self.get_size_t_array(self.fastbins_addr, self.num_fastbins)
+    @property
+    def top(self):
+        return self.get_size_t_pointer(self.top_addr)
+    @property
+    def last_remainder(self):
+        return self.get_size_t_pointer(self.last_remainder_addr)
+    @property
+    def bins(self):
+        return self.get_size_t_array(self.bins_addr, self.num_bins)
+    @property
+    def next(self):
+        return self.get_size_t_pointer(self.next_addr)
+    @property
+    def next_free(self):
+        return self.get_size_t_pointer(self.next_free_addr)
+    @property
+    def system_mem(self):
+        return self.get_size_t(self.system_mem_addr)
+
+    def get_size_t(self, addr):
+        return dereference(addr).cast(self.size_t)
+    def get_size_t_pointer(self, addr):
+        size_t_pointer = self.size_t.pointer()
+        return dereference(addr).cast(size_t_pointer)
+    def get_size_t_array(self, addr, length):
+        size_t_array = self.size_t.array(length)
+        return dereference(addr).cast(size_t_array)
+
+    def __getitem__(self, item):
+        return getattr(self, item)
+
+@lru_cache()
+def locate_main_arena_address(addr):
+    try:
+        return to_unsigned_long(gdb.parse_and_eval("&{}".format(addr)))
+    except gdb.error:
+        pass
```

Comment:
I think that if the address is provided but then it fails we should raise an exception. But of course if it isn't provided then falling back to the trick below makes sense

```python
if addr is not None:
    return <use address> # This might raise an Exception

return <hook trick>
```

Reply:
`addr` is always provided though (`GlibcArena` requires it), it is just normally `main_arena`. Have to have a think